File: py-code/技巧/二分查找/LC_2563.py

# 2563. 统计公平数对的数目
# 第 332 场周赛
# Q2
# 1721
# 相关标签
# 相关企业
# 提示
# 给你一个下标从 0 开始、长度为 n 的整数数组 nums ，和两个整数 lower 和 upper ，返回 公平数对的数目 。

# 如果 (i, j) 数对满足以下情况，则认为它是一个 公平数对 ：

# 0 <= i < j < n，且
# lower <= nums[i] + nums[j] <= upper


from bisect import *
from typing import List


class Solution:
    def countFairPairs(self, nums: List[int], lower: int, upper: int) -> int:
        ans = 0
        nums.sort()
        for j, x in enumerate(nums):
            l = bisect_left(nums, lower - x, 0, j)
            r = bisect_right(nums, upper - x, 0, j)
            ans += r - l
        return ans

File: leetcode/src/com/fqh/二分查找/LC_2563.java
package com.fqh.二分查找;

import java.util.Arrays;

/**
 * @author ikun
 * @version v1.0.0
 * @since 2024/4/1 21:15
 **/
public class LC_2563 {

//    https://leetcode.cn/problems/count-the-number-of-fair-pairs/description/

    public long countFairPairs(int[] nums, int lower, int upper) {
        Arrays.sort(nums);
        long ans = 0;
        for (int i = 0; i < nums.length; i++) {
            int l = bisectLeft(nums, i + 1, lower - nums[i]);
            int r = bisectRight(nums, i + 1, upper - nums[i]);
            if (l == -1 || r == -1) continue;
            ans += r - l;
        }
        return ans;
    }

    public static int bisectLeft(int[] a, int start, int target) {
        int left = start, right = a.length - 1;
        while (left <= right) {
            int mid = (left + right) >> 1;
            if (a[mid] < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }

    public static int bisectRight(int[] a, int start, int target) {
        int left = start, right = a.length - 1;
        while (left <= right) {
            int mid = (left + right) >> 1;
            if (a[mid] <= target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }
        return left;
    }
}
